### PR TITLE
Use a checkable combobox for fields configuration flags

### PR DESCRIPTION
**Pull request by 3nids**
image

this also migrates the WMS/WFS exclusion to configuration flags

### CODE REVIEW

File: python/core/auto_generated/qgsfield.sip.in [hunk 1]
```diff
@@ -291,6 +291,13 @@ Sets the alias for the field (the friendly displayed name of the field ).
 Formats string for display
 %End
 
+    static QString readableConfigurationFlag( ConfigurationFlag flag );
+%Docstring
+Returns the reabable and translated value of the configuration flag
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsfield.sip.in:296]:
```suggestion
Returns a readable and translated value of the configuration flag
```

File: python/core/auto_generated/qgsvectorlayer.sip.in [hunk 2]
```diff
@@ -1968,24 +1968,32 @@ Convenience function that returns the attribute alias if defined or the field na
 Returns a map of field name to attribute alias
 %End
 
-    QSet<QString> excludeAttributesWms() const;
+ QSet<QString> excludeAttributesWms() /Deprecated/ const;
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsvectorlayer.sip.in:1971]: Yay!!

File: python/core/auto_generated/qgsvectorlayer.sip.in [hunk 3]
```diff
@@ -1968,24 +1968,32 @@ Convenience function that returns the attribute alias if defined or the field na
...
 %Docstring
 A set of attributes that are not advertised in WMS requests with QGIS server.
+
+.. deprecated:: QGIS 3.16
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsvectorlayer.sip.in:1975]: You should add a note here about what the replacement is

File: src/core/qgsfield.cpp [hunk 4]
```diff
@@ -346,6 +346,23 @@ QString QgsField::displayString( const QVariant &v ) const
...
+QString QgsField::readableConfigurationFlag( QgsField::ConfigurationFlag flag )
+{
+  switch ( flag )
+  {
+    case ConfigurationFlag::None:
+      return QObject::tr( "None" );
+    case ConfigurationFlag::Searchable:
+      return QObject::tr( "Searchable" );
+    case ConfigurationFlag::Wms:
+      return QStringLiteral( "WMS" );
```
Comment (nyalldawson) [inline src/core/qgsfield.cpp:358]:
```suggestion
      return QStringLiteral( "Expose via WMS" );
```

File: src/core/qgsfield.cpp [hunk 5]
```diff
@@ -346,6 +346,23 @@ QString QgsField::displayString( const QVariant &v ) const
...
+    case ConfigurationFlag::Wfs:
+      return QStringLiteral( "WFS" );
```
Comment (nyalldawson) [inline src/core/qgsfield.cpp:360]:
```suggestion
      return QStringLiteral( "Expose via WFS" );
```

File: src/core/qgsfield.h [hunk 6]
```diff
@@ -80,8 +80,10 @@ class CORE_EXPORT QgsField
 #endif
     {
       None = 0, //!< No flag is defined
-      Searchable = 0x1, //!< Defines if the field is searchable (used in the locator search for instance)
-      DefaultFlags = Searchable, //!< Default set of flags for a field
+      Searchable = 1 << 1, //!< Defines if the field is searchable (used in the locator search for instance)
+      Wms = 1 << 2, //!< Fields is available if layer is served as WMS
```
Comment (nyalldawson) [inline src/core/qgsfield.h:84]:
```suggestion
      ExposeViaWms = 1 << 2, //!< Fields is available if layer is served as WMS from QGIS server
```

File: src/core/qgsfield.h [hunk 7]
```diff
@@ -80,8 +80,10 @@ class CORE_EXPORT QgsField
...
+      Wfs = 1 << 3, //!< Fields is available if layer is served as WFS
```
Comment (nyalldawson) [inline src/core/qgsfield.h:85]:
```suggestion
      ExposeViaWfs = 1 << 3, //!< Fields is available if layer is served as WFS from QGIS server
```

**Comment by 3nids** [comment]
@nyalldawson thanks for the review

**Pull request merged**